### PR TITLE
Separate schedule_upcall number and values through Rust types

### PR DESCRIPTION
**Pull request by alexandruradovici**
### Pull Request Overview

This pull request groups the values returned by `schedule_upcall` into a tuple to prevent wrongly placing a value instead of the upcall number.

```rust
pub fn schedule_upcall(
        &self,
        subscribe_num: usize,
        r: (usize, usize, usize),
    ) -> Result<(), UpcallError> 
```

This tries to use Rust's type system to clearly separate the upcall number and the values. The `schedule_upcall` takes four `usize` parameters, the first one being the upcall number and the following three the values returned to the application. As all four parameters have the same type, it is easy to forget that the first parameter is the upcall number, not the first value. The compiler has no way of protecting against this error. 

This is inspired from the error within #2698.

There should be no impact in size or performance.

### Testing Strategy

N/A

### TODO or Help Wanted

N/A

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

**Review by alistair23 — DISMISSED** [review]
I like this! I agree the current setup is confusing and easy to get wrong.

The other option we could do is an `enum` for the first argument, that way it isn't just a simple number.

**Comment by alexandruradovici** [comment]
> I like this! I agree the current setup is confusing and easy to get wrong.
> 
> The other option we could do is an `enum` for the first argument, that way it isn't just a simple number.

How do you see this `enum`? Somthing that implements a certain trait?

**Comment by lschuermann** [comment]
> How do you see this `enum`? Somthing that implements a certain trait?

I believe @alistair23 is referring to something akin to `capsules::driver::NUM` (if not that specifically):
https://github.com/tock/tock/blob/92eafdb491c3e6d14e6d798ccd8b0af959619cec/capsules/src/driver.rs#L6-L9

That being that, I don't think that would be a great idea. Driver numbers can be well-known, but it's also strictly possible and desired for developers to create their own numbers for custom drivers, or multiple instances of the same one. Thus, using a number seems much more sensible.

File: doc/Porting_v1_Capsules_to_v2.md [hunk 1]
```diff
@@ -311,7 +311,7 @@ replaced with code like:
 
 ```rust
 self.apps.enter(appid, |app, upcalls| {
-    upcalls.schedule_upcall(upcall_number, r0, r1, r2);
+    upcalls.schedule_upcall(upcall_number,(r0, r1, r2));
```
Comment (bradjc) [inline doc/Porting_v1_Capsules_to_v2.md:314]:
```suggestion
    upcalls.schedule_upcall(upcall_number, (r0, r1, r2));
```

**Review by bradjc — APPROVED** [review]

**Review by bradjc — APPROVED** [review]

**Comment by alistair23** [comment]
As @lschuermann said something like this:

```rust
pub enum CallbackNumber {
    ZERO = 0,
    ONE = 1,
    TWO = 2,
    THREE = 3,
    FOUR = 4,
    FIVE = 5,
    SIX = 6,
}
```

That way it's more obvious what is going on

**Review by ppannuto — APPROVED** [review]
I think this is good to go as-is. The `enum` is an interesting idea, but I'm not quite convinced the ergonomic overhead is going to be worth it (especially given that it also adds imports everywhere). I think I would also be worried as a user about in what ways `ZERO` is or is not just `0`, which might confuse more than help.

**Review by lschuermann — APPROVED** [review]
I agree with @ppannuto. The enum is indeed an interesting idea, but the fact that it requires imports, is another thing we keep around and maintain or generate through some macro, and is non-standard thus might confuse people makes me think that a simple numeric seems like the simpler option.

I've taken the liberty to transfer this out of the RFC phase, given it has been discussed on today's call.

**Review by brghena — APPROVED** [review]

**Pull request merged**